core/arch/atomic_exports.h

```diff
@@ -730,24 +730,24 @@ atomic_dec_becomes_zero(volatile int *var)
 #        define SET_IF_NOT_ZERO(flag) SET_FLAG(ne, flag)
 #        define SET_IF_NOT_LESS(flag) SET_FLAG(ge, flag)
 #    elif defined(DR_HOST_RISCV64)
-#        define ATOMIC_1BYTE_READ(addr_src, addr_res)               \
-            do {                                                    \
-                /* Fence for acquire semantics. */                  \
-                __asm__ __volatile__("lbu %0, 0(%1)\n\t"            \
-                                     "fence iorw,iorw"              \
-                                     :                              \
-                                     : "r"(addr_src), "r"(addr_res) \
-                                     : "memory");                   \
+#        define ATOMIC_1BYTE_READ(addr_src, res)         \
```

Comment:
RISC-V's load-reserved (LR) and store-conditional (SC) instructions are tricky, they cannot be used with normal memory access instructions. E.g. it's invalid to do something like this:

```
LR.W ...
LW   ...
SC.W ...
```

IIUC, something like this is totally fine in ARM.

Reply:
From riscv-spec-20191213.pdf 8.3 Eventual Success of Store-Conditional Instructions

> An LR/SC sequence begins with an LR instruction and ends with an SC instruction. The dynamic code executed between the LR and SC instructions can only contain instructions from the base “I” instruction set, excluding loads, stores, backward jumps, taken backward branches, JALR, FENCE, FENCE.I, and SYSTEM instructions. If the “C” extension is supported, then compressed forms of the aforementioned “I” instructions are also permitted.

Reply:
The destination of these ATOMIC_xBYTE_READ macors would be a local memory location and the source would be a shared memory location. So only the read part needs to be "atomic" (that is, it should follow acquire-release semantics and (for multiple bytes) be read together). It doesn't make sense to write "atomically" to a local location.

So we can probably read into a temporary reg (like we're doing now), and then have a separate regular write to memory at addr_res.

@derekbruening to verify my understanding is correct.

Reply:
Yes, writing the result for the caller is a separate, local step.  No need to add the complexity of changing the signature just for one arch.

"Atomic" for DR means untorn and acquire-release semantics.
Generally these macros should do the same thing the compiler would for `res = std::atomic<X>::load(std::memory_order_acquire)` (or `store` or `fetch_add` with `std::memory_order_release`).  
I would suggest directly observing what a compiler does for each of these to avoid making a mistake trying to hand-construct a sequence which can be error-prone.  Xref #4264 on using the compiler in the first place now that modern compilers have support -- but that would change the types and we'd need every compiler on board.